### PR TITLE
feat: includeMirrors

### LINKED ISSUE #16436: support includeMirrors
### What would you like Renovate to be able to do?

Original discussion https://github.com/renovatebot/renovate/discussions/16327

### If you have any ideas on how this should be implemented, please tell us here.

support includeMirrors to avoid "Repository is a mirror - skipping" 

### Is this a feature you are interested in implementing yourself?

No

### PR DESCRIPTION
**Pull request by Rainman73**
## Changes

Add support for 'includeMirrors' configuration option to allow Renovate to operate on (gitlab) repositories that are mirrors.

## Context

Closes #16436

## Documentation

- [x] I have updated the documentation, or
- [ ] No documentation update is required

## How I've tested my work 

I have verified these changes via:

- [ ] Code inspection only, or
- [x] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

**Comment by Rainman73** [comment]
As you can probably tell, I'm not a JS/TS programmer. I'm open to suggestions for more idiomatic solutions.

**Review by rarkins — DISMISSED** [review]

**Review by rarkins — COMMENTED** [review]
Does it need to be globalOnly or can it work with repo config?

File: lib/modules/platform/gitlab/index.ts [hunk 1]
```diff
@@ -263,11 +264,13 @@ export async function initRepo({
   ignorePrAuthor,
   gitUrl,
   endpoint,
+  includeMirrors,
 }: RepoParams): Promise<RepoResult> {
   config = {} as any;
   config.repository = urlEscape(repository);
   config.cloneSubmodules = cloneSubmodules;
   config.ignorePrAuthor = ignorePrAuthor;
+  config.includeMirrors = includeMirrors;
```
Comment (viceice) [inline lib/modules/platform/gitlab/index.ts:273]:
```suggestion
```
i don't think we need to store that option

File: lib/modules/platform/gitlab/index.ts [hunk 2]
```diff
@@ -280,7 +283,7 @@ export async function initRepo({
       );
       throw new Error(REPOSITORY_ARCHIVED);
     }
-    if (res.body.mirror) {
+    if (res.body.mirror && !config.includeMirrors) {
```
Comment (viceice) [inline lib/modules/platform/gitlab/index.ts:286]:
```suggestion
    if (res.body.mirror && includeMirrors !== true) {
```
more explicit

File: lib/modules/platform/gitlab/index.ts [hunk 3]
```diff
@@ -72,6 +72,7 @@ let config: {
   cloneSubmodules: boolean | undefined;
   ignorePrAuthor: boolean | undefined;
   squash: boolean;
+  includeMirrors: boolean | undefined;
```
Comment (viceice) [inline lib/modules/platform/gitlab/index.ts:75]:
```suggestion
```
not needed, see below

**Review by viceice — COMMENTED** [review]
mirror repos are still filtered on autodiscover, check `getRepos` function

**Comment by Rainman73** [comment]
> mirror repos are still filtered on autodiscover, check `getRepos` function



> mirror repos are still filtered on autodiscover, check `getRepos` function

Good point, the `.filter()` needs to be updated there, but how do I get the value of the `includeMirrors` setting there? (Sorry for the newb question)

File: docs/usage/configuration-options.md [hunk 4]
```diff
@@ -1602,6 +1602,12 @@ If you need to force permanent unstable updates for a package, you can add a pac
 
 Also check out the `followTag` configuration option above if you wish Renovate to keep you pinned to a particular release tag.
 
+## includeMirrors
```
Comment (rarkins) [inline docs/usage/configuration-options.md:1605]: Move to self-hosted file

File: lib/config/options/index.ts [hunk 5]
```diff
@@ -442,6 +442,14 @@ const options: RenovateOptions[] = [
     allowedValues: ['auto', 'enabled', 'disabled'],
     default: 'auto',
   },
+  {
+    name: 'includeMirrors',
+    description:
+      'Whether to process repositories that are mirrors. By default, repositories that are mirrors are skipped.',
+    type: 'boolean',
+    default: false,
+    supportedPlatforms: ['gitlab'],
```
Comment (rarkins) [inline lib/config/options/index.ts:451]:
```suggestion
    supportedPlatforms: ['gitlab'],
    globalOnly: true,
```

**Review by rarkins — CHANGES_REQUESTED** [review]
Make global only

**Comment by rarkins** [comment]
Can you confirm that it works with --autodiscover flag?

**Comment by Rainman73** [comment]
> Can you confirm that it works with --autodiscover flag?

Actually I'm pretty sure it does *not* work with autodiscover because of the issue mentioned by viceice above. 

How do I get the value of the `includeMirrors` setting in the `getRepos` function in `lib/modules/platform/gitlab/index.ts` ?

**Comment by rarkins** [comment]
@viceice I don't see any mirror filtering here: https://github.com/renovatebot/renovate/blob/81eb17aacac37dff6fa64def330f8fb0b428f4a3/lib/modules/platform/gitlab/index.ts#L146-L174

**Comment by Rainman73** [comment]
> @viceice I don't see any mirror filtering here:

Not viceice, but I think the filtering is being done on line 168:
https://github.com/renovatebot/renovate/blob/81eb17aacac37dff6fa64def330f8fb0b428f4a3/lib/modules/platform/gitlab/index.ts#L167-L169

**Comment by rarkins** [comment]
Yes you're right, I missed that! So that line needs modification too

**Review by rarkins — CHANGES_REQUESTED** [review]
Needs update to `getRepos()`

**Comment by Rainman73** [comment]
> Needs update to `getRepos()`

I updated `getRepos()` to use `GlobalConfig` and added a testcase, but when I set `includeMirrors` to true in the tests, the changes to GlobalConfig seem to be ignored. Should the `GlobalConfig` be mocked for the tests?

**Review by rarkins — CHANGES_REQUESTED** [review]
Please modify `AutodiscoverConfig` instead of using GlobalConfig and then pass the param here: https://github.com/renovatebot/renovate/blob/440ea597fcab0fe687882f8170e72a57e4930f89/lib/workers/global/autodiscover.ts#L37-L40

**Comment by Rainman73** [comment]
As far as I can tell, I have made all requested changes, and the newly added tests now work correctly. Thanks for the patience!

File: test/setup.ts [hunk 6]
```diff
@@ -151,7 +151,7 @@ declare global {
     type MockInstance<T, Y extends any[]> = JestMockInstance<(...args: Y) => T>;
 
     interface Mock<T = any, Y extends any[] = any>
-      extends Function, // eslint-disable-line @typescript-eslint/ban-types
+      extends Function,
```
Comment (rarkins) [inline test/setup.ts:154]: Unrelated change

File: docs/usage/self-hosted-configuration.md [hunk 7]
```diff
@@ -482,6 +482,12 @@ Use the `extends` field instead of this if, for example, you need the ability fo
     When Renovate resolves `globalExtends` it does not fully process the configuration.
     This means that Renovate does not have the authentication it needs to fetch private things.
 
+## includeMirrors
+
+By default, Renovate does not operate on repositories that are mirrors.
```
Comment (rarkins) [inline docs/usage/self-hosted-configuration.md:487]:
```suggestion
By default, Renovate does not autodiscover repositories that are mirrors.
```

File: lib/modules/platform/gitlab/index.ts [hunk 8]
```diff
@@ -267,6 +270,7 @@ export async function initRepo({
   ignorePrAuthor,
   gitUrl,
   endpoint,
+  includeMirrors,
```
Comment (rarkins) [inline lib/modules/platform/gitlab/index.ts:273]:
```suggestion
```

File: lib/modules/platform/gitlab/index.ts [hunk 9]
```diff
@@ -284,7 +288,8 @@ export async function initRepo({
       );
       throw new Error(REPOSITORY_ARCHIVED);
     }
-    if (res.body.mirror) {
+
+    if (res.body.mirror && includeMirrors !== true) {
```
Comment (rarkins) [inline lib/modules/platform/gitlab/index.ts:292]: @viceice What if we remove this if altogether?

1. If it's been autodiscovered then we know the includeMirrors flag must be true already
2. If the user has run it directly, they clearly want to include it

**Review by rarkins — CHANGES_REQUESTED** [review]

File: test/setup.ts (hunk 6 above)
Comment (Rainman73) [reply test/setup.ts:154]: In my (docker/vs code) dev env `yarn test` stumbles over this because of
 `154:25  error  Unused eslint-disable directive (no problems were reported from '@typescript-eslint/ban-types')`

Am I doing something wrong, or should the eslint directive be reinstated?

Comment (rarkins) [reply test/setup.ts:154]: Maybe you needed to run `yarn` again to get the latest dependencies

**Comment by rarkins** [comment]
Global only.

File: lib/modules/platform/gitlab/index.ts [hunk 10]
```diff
@@ -168,8 +169,10 @@ export async function getRepos(config?: AutodiscoverConfig): Promise<string[]> {
       paginate: true,
     });
     logger.debug(`Discovered ${res.body.length} project(s)`);
+    const includeMirrors = config?.includeMirrors === true;
+
```
Comment (rarkins) [inline lib/modules/platform/gitlab/index.ts:173]:
```suggestion
```

File: lib/modules/platform/gitlab/index.ts [hunk 11]
```diff
@@ -168,8 +169,10 @@ export async function getRepos(config?: AutodiscoverConfig): Promise<string[]> {
...
     return res.body
-      .filter((repo) => !repo.mirror)
+      .filter((repo) => includeMirrors || !repo.mirror)
```
Comment (rarkins) [inline lib/modules/platform/gitlab/index.ts:175]:
```suggestion
      .filter((repo) => config?.includeMirrors || !repo.mirror)
```

File: lib/modules/platform/gitlab/index.ts (hunk 9 above)
Comment (rarkins) [reply lib/modules/platform/gitlab/index.ts:292]: Let's do it later

**Review by rarkins — CHANGES_REQUESTED** [review]

**Review by rarkins — DISMISSED** [review]

File: lib/modules/platform/gitlab/index.ts [hunk 12]
```diff
@@ -169,7 +170,7 @@ export async function getRepos(config?: AutodiscoverConfig): Promise<string[]> {
     });
     logger.debug(`Discovered ${res.body.length} project(s)`);
     return res.body
-      .filter((repo) => !repo.mirror)
+      .filter((repo) => config?.includeMirrors || !repo.mirror) // eslint-disable-line @typescript-eslint/prefer-nullish-coalescing
```
Comment (viceice) [inline lib/modules/platform/gitlab/index.ts:173]: don't suppress this lint error

File: lib/modules/platform/gitlab/index.ts [hunk 13]
```diff
@@ -72,6 +72,7 @@ let config: {
   cloneSubmodules: boolean | undefined;
   ignorePrAuthor: boolean | undefined;
   squash: boolean;
+  includeMirrors: boolean;
```
Comment (viceice) [inline lib/modules/platform/gitlab/index.ts:75]:
```suggestion
```
seems to be unused

File: lib/modules/platform/gitlab/index.ts (hunk 12 above)
Comment (Rainman73) [reply lib/modules/platform/gitlab/index.ts:173]: Without suppressing the lint error the build fails, which is why I used  
`const includeMirrors = config?.includeMirrors === true;` before, which @rarkins did not like.
What is the neatest way to fix this here?

File: lib/config/global.ts [hunk 14]
```diff
@@ -28,6 +28,7 @@ export class GlobalConfig {
     'gitTimeout',
     'platform',
     'endpoint',
+    'includeMirrors',
```
[the remark this answers is not in the document]
Comment (Rainman73) [reply lib/config/global.ts]: You're right, I forgot to remove this when undoing the GlobalConfig changes

**Comment by rarkins** [comment]
@Rainman73 please don't push from main branch unless you're making other changes. Each push resets the test approvals

**Review by rarkins — APPROVED** [review]
